sdk/data/azcosmos/cosmos_session_policy.go

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package azcosmos
+
+import (
+	"net/http"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+)
+
+type sessionPolicy struct {
+	sc interface {
+		GetSessionToken(resourceAddress string) string
+		SetSessionToken(resourceAddress string, containerRid string, sessionToken string)
+		ClearSessionToken(resourceAddress string)
+	}
+}
+
+func (p *sessionPolicy) Do(req *policy.Request) (*http.Response, error) {
+	o := pipelineRequestOptions{}
+	if !req.OperationValue(&o) || req.Raw().Header.Get(cosmosHeaderSessionToken) != "" {
+		return req.Next()
+	}
+
+	// Reads: use session token if available
+	if !o.isWriteOperation {
+		sessionToken := p.sc.GetSessionToken(o.resourceAddress)
+		if sessionToken != "" {
+			req.Raw().Header.Set(cosmosHeaderSessionToken, sessionToken)
+		}
+	}
+
+	response, err := req.Next()
+	if err != nil || !(response.StatusCode >= 200 && response.StatusCode < 300) {
+		// Allow potential retry attempt without session token
+		p.sc.ClearSessionToken(o.resourceAddress)
+		return response, err
+	}
+
+	// Successful Writes: cache session token
+	if o.isWriteOperation && response.StatusCode >= 200 && response.StatusCode < 300 {
+		sessionToken := response.Header.Get(cosmosHeaderSessionToken)
+		containerPath := response.Header.Get(cosmosHeaderAltContentPath)
+		containerRid := response.Header.Get(cosmosHeaderContentPath)
+
+		// We currently expect the container path and RID headers to be present
+		// (rather than falling back to o.resourceAddress and parsing _self)
```

Comment:
I don't know if this is a safe assumption - other clients (Python, JS) have various behavior in terms of expecting both `X-Ms-Alt-Content-Path` and `X-Ms-Content-Path`. And the latter isn't documented at https://learn.microsoft.com/en-us/rest/api/cosmos-db/common-cosmosdb-rest-response-headers

It _seems_ like both are always set (at least for name-based requests) when a response includes a session token, but I wonder if this should be (falling back to?) getting the container rid by parsing `_self` and getting the container path by falling back to o.resourceAddress.